src/Compilers/CSharp/Portable/Lowering/ClosureConversion/LambdaCapturedVariable.cs

```diff
@@ -39,21 +41,38 @@ private LambdaCapturedVariable(SynthesizedClosureEnvironment frame, TypeWithAnno
         public SynthesizedClosureEnvironment Frame
             => (SynthesizedClosureEnvironment)ContainingType;
 
+#nullable enable
+
         public static LambdaCapturedVariable Create(SynthesizedClosureEnvironment frame, Symbol captured, ref int uniqueId)
         {
             Debug.Assert(captured is LocalSymbol || captured is ParameterSymbol);
 
             string fieldName = GetCapturedVariableFieldName(captured, ref uniqueId);
             TypeSymbol type = GetCapturedVariableFieldType(frame, captured);
-            return new LambdaCapturedVariable(frame, TypeWithAnnotations.Create(type), fieldName, IsThis(captured));
+            bool isThis = IsThis(captured, out ParameterSymbol? parameter);
+            return parameter is not null && !isThis ?
+                new LambdaCapturedVariableForRegularParameter(frame, TypeWithAnnotations.Create(type), fieldName, parameter) :
+                new LambdaCapturedVariable(frame, TypeWithAnnotations.Create(type), fieldName, isThis);
         }
 
         private static bool IsThis(Symbol captured)
```

Comment:
>IsThis [](http://example.com/codeflow?start=28&length=6)

It looks like this overload is only used in one location, in `GetCapturedVariableFieldName()` below. Consider inlining the call there and removing this overload.